in version 2.3.1 and 2.4.0 Vault MicroProfile Config Source stopped working with OIDC

### Describe the bug

in version 2.3.1 and 2.4.0 Vault MicroProfile Config Source stopped working to use vault with OIDC (Vault MicroProfile Config Source)

### Expected behavior

quarkus fetched vault kv values first and then try verificted dependent properties. In version 2.1.2 works fine, but in 2.3.1 and 2.4.0 does not work

### Actual behavior

on startup throw error:

```
2021-10-27 15:40:13,605 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor#prepareOidcDevConsole threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.client-id in property quarkus.oidc.client-id
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:139)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:821)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:418)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:276)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:67)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:92)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:441)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:65)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor#prepareOidcDevConsole threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.client-id in property quarkus.oidc.client-id
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:139)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:821)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.builder.Execution.run(Execution.java:116)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:161)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:416)
	... 9 more
Caused by: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.client-id in property quarkus.oidc.client-id
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:139)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:821)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Linux localhost 5.14.11-1-default #1 SMP Sun Oct 10 08:34:34 UTC 2021 (834dddd) x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.12" 2021-07-20 LTS OpenJDK Runtime Environment Corretto-11.0.12.7.1 (build 11.0.12+7-LTS) OpenJDK 64-Bit Server VM Corretto-11.0.12.7.1 (build 11.0.12+7-LTS, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.4.0.FINAL

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Hi @radcortez, @vsevel: please note that this error is similar to old #14707 but only with OIDC.

--- a/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/AbstractDevConsoleProcessor.java
+++ b/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/AbstractDevConsoleProcessor.java
@@ -13,8 +13,6 @@ public abstract class AbstractDevConsoleProcessor {
             String oidcProviderName,
             String oidcApplicationType,
             String oidcGrantType,
-            String clientId,
-            String clientSecret,
             String authorizationUrl,
             String tokenUrl,
             String logoutUrl,
@@ -24,9 +22,6 @@ public abstract class AbstractDevConsoleProcessor {
         }
         devConsoleTemplate.produce(new DevConsoleTemplateInfoBuildItem("oidcApplicationType", oidcApplicationType));
         devConsoleTemplate.produce(new DevConsoleTemplateInfoBuildItem("oidcGrantType", oidcGrantType));
-
-        devConsoleTemplate.produce(new DevConsoleTemplateInfoBuildItem("clientId", clientId));
-        devConsoleTemplate.produce(new DevConsoleTemplateInfoBuildItem("clientSecret", clientSecret));
 
         devConsoleTemplate.produce(new DevConsoleTemplateInfoBuildItem("authorizationUrl", authorizationUrl));
         devConsoleTemplate.produce(new DevConsoleTemplateInfoBuildItem("tokenUrl", tokenUrl));

--- a/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevConsoleProcessor.java
+++ b/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevConsoleProcessor.java
@@ -14,9 +14,11 @@ import io.quarkus.deployment.annotations.Consume;
 import io.quarkus.deployment.builditem.CuratedApplicationShutdownBuildItem;
 import io.quarkus.deployment.builditem.RuntimeConfigSetupCompleteBuildItem;
 import io.quarkus.devconsole.spi.DevConsoleRouteBuildItem;
+import io.quarkus.devconsole.spi.DevConsoleRuntimeTemplateInfoBuildItem;
 import io.quarkus.devconsole.spi.DevConsoleTemplateInfoBuildItem;
 import io.quarkus.oidc.common.runtime.OidcConstants;
 import io.quarkus.oidc.deployment.OidcBuildTimeConfig;
+import io.quarkus.oidc.runtime.OidcConfigPropertySupplier;
 import io.quarkus.runtime.configuration.ConfigUtils;
 import io.vertx.core.Vertx;
 import io.vertx.core.http.HttpHeaders;
@@ -47,7 +49,8 @@ public class OidcDevConsoleProcessor extends AbstractDevConsoleProcessor {
 
     @BuildStep(onlyIf = IsDevelopment.class)
     @Consume(RuntimeConfigSetupCompleteBuildItem.class)
-    void prepareOidcDevConsole(BuildProducer<DevConsoleTemplateInfoBuildItem> console,
+    void prepareOidcDevConsole(BuildProducer<DevConsoleTemplateInfoBuildItem> devConsoleInfo,
+            BuildProducer<DevConsoleRuntimeTemplateInfoBuildItem> devConsoleRuntimeInfo,
             CuratedApplicationShutdownBuildItem closeBuildItem,
             BuildProducer<DevConsoleRouteBuildItem> devConsoleRoute,
             Capabilities capabilities) {
@@ -79,20 +82,25 @@ public class OidcDevConsoleProcessor extends AbstractDevConsoleProcessor {
             }
             String providerName = tryToGetProviderName(authServerUrl);
             if (KEYCLOAK.equals(providerName)) {
-                console.produce(new DevConsoleTemplateInfoBuildItem("keycloakAdminUrl",
+                devConsoleInfo.produce(new DevConsoleTemplateInfoBuildItem("keycloakAdminUrl",
                         authServerUrl.substring(0, authServerUrl.indexOf("/realms/"))));
             }
             produceDevConsoleTemplateItems(capabilities,
-                    console,
+                    devConsoleInfo,
                     providerName,
                     getApplicationType(),
                     oidcConfig.devui.grant.type.isPresent() ? oidcConfig.devui.grant.type.get().getGrantType() : "code",
-                    getConfigProperty(CLIENT_ID_CONFIG_KEY),
-                    getClientSecret(),
                     metadata.getString("authorization_endpoint"),
                     metadata.getString("token_endpoint"),
                     metadata.getString("end_session_endpoint"),
                     metadata.containsKey("introspection_endpoint"));
+
+            devConsoleRuntimeInfo.produce(
+                    new DevConsoleRuntimeTemplateInfoBuildItem("clientId",
+                            new OidcConfigPropertySupplier(CLIENT_ID_CONFIG_KEY)));
+            devConsoleRuntimeInfo.produce(
+                    new DevConsoleRuntimeTemplateInfoBuildItem("clientSecret",
+                            new OidcConfigPropertySupplier(CLIENT_SECRET_CONFIG_KEY, "")));
 
             Duration webClientTimeout = oidcConfig.devui.webClienTimeout.isPresent() ? oidcConfig.devui.webClienTimeout.get()
                     : Duration.ofSeconds(4);
@@ -150,10 +158,6 @@ public class OidcDevConsoleProcessor extends AbstractDevConsoleProcessor {
 
     private static boolean isClientIdSet() {
         return ConfigUtils.isPropertyPresent(CLIENT_ID_CONFIG_KEY);
-    }
-
-    private static String getClientSecret() {
-        return ConfigProvider.getConfig().getOptionalValue(CLIENT_SECRET_CONFIG_KEY, String.class).orElse("");
     }
 
     private static boolean isAuthServerUrlSet() {

--- a/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakDevConsoleProcessor.java
+++ b/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/keycloak/KeycloakDevConsoleProcessor.java
@@ -11,43 +11,54 @@ import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Consume;
 import io.quarkus.deployment.builditem.RuntimeConfigSetupCompleteBuildItem;
 import io.quarkus.devconsole.spi.DevConsoleRouteBuildItem;
+import io.quarkus.devconsole.spi.DevConsoleRuntimeTemplateInfoBuildItem;
 import io.quarkus.devconsole.spi.DevConsoleTemplateInfoBuildItem;
 import io.quarkus.oidc.deployment.OidcBuildTimeConfig;
 import io.quarkus.oidc.deployment.devservices.AbstractDevConsoleProcessor;
 import io.quarkus.oidc.deployment.devservices.OidcAuthorizationCodePostHandler;
 import io.quarkus.oidc.deployment.devservices.OidcPasswordClientCredHandler;
 import io.quarkus.oidc.deployment.devservices.OidcTestServiceHandler;
+import io.quarkus.oidc.runtime.OidcConfigPropertySupplier;
 
 public class KeycloakDevConsoleProcessor extends AbstractDevConsoleProcessor {
+
+    private static final String CONFIG_PREFIX = "quarkus.oidc.";
+    private static final String CLIENT_ID_CONFIG_KEY = CONFIG_PREFIX + "client-id";
+    private static final String CLIENT_SECRET_CONFIG_KEY = CONFIG_PREFIX + "credentials.secret";
 
     KeycloakBuildTimeConfig keycloakConfig;
     OidcBuildTimeConfig oidcConfig;
 
     @BuildStep(onlyIf = IsDevelopment.class)
     @Consume(RuntimeConfigSetupCompleteBuildItem.class)
-    public void setConfigProperties(BuildProducer<DevConsoleTemplateInfoBuildItem> console,
+    public void setConfigProperties(BuildProducer<DevConsoleTemplateInfoBuildItem> devConsoleInfo,
+            BuildProducer<DevConsoleRuntimeTemplateInfoBuildItem> devConsoleRuntimeInfo,
             Optional<KeycloakDevServicesConfigBuildItem> configProps,
             Capabilities capabilities) {
         if (configProps.isPresent() && configProps.get().getProperties().containsKey("keycloak.url")) {
             String keycloakUrl = (String) configProps.get().getProperties().get("keycloak.url");
             String realmUrl = keycloakUrl + "/realms/" + configProps.get().getProperties().get("keycloak.realm");
 
-            console.produce(new DevConsoleTemplateInfoBuildItem("keycloakAdminUrl", keycloakUrl));
-            console.produce(
+            devConsoleInfo.produce(new DevConsoleTemplateInfoBuildItem("keycloakAdminUrl", keycloakUrl));
+            devConsoleInfo.produce(
                     new DevConsoleTemplateInfoBuildItem("keycloakUsers", configProps.get().getProperties().get("oidc.users")));
 
             produceDevConsoleTemplateItems(capabilities,
-                    console,
+                    devConsoleInfo,
                     "Keycloak",
                     (String) configProps.get().getProperties().get("quarkus.oidc.application-type"),
                     oidcConfig.devui.grant.type.isPresent() ? oidcConfig.devui.grant.type.get().getGrantType()
                             : keycloakConfig.devservices.grant.type.getGrantType(),
-                    (String) configProps.get().getProperties().get("quarkus.oidc.client-id"),
-                    (String) configProps.get().getProperties().get("quarkus.oidc.credentials.secret"),
                     realmUrl + "/protocol/openid-connect/auth",
                     realmUrl + "/protocol/openid-connect/token",
                     realmUrl + "/protocol/openid-connect/logout",
                     true);
+            devConsoleRuntimeInfo.produce(
+                    new DevConsoleRuntimeTemplateInfoBuildItem("clientId",
+                            new OidcConfigPropertySupplier(CLIENT_ID_CONFIG_KEY)));
+            devConsoleRuntimeInfo.produce(
+                    new DevConsoleRuntimeTemplateInfoBuildItem("clientSecret",
+                            new OidcConfigPropertySupplier(CLIENT_SECRET_CONFIG_KEY, "")));
 
         }
     }

--- a/extensions/oidc/deployment/src/test/java/io/quarkus/oidc/test/CodeFlowDevModeTestCase.java
+++ b/extensions/oidc/deployment/src/test/java/io/quarkus/oidc/test/CodeFlowDevModeTestCase.java
@@ -2,12 +2,12 @@ package io.quarkus.oidc.test;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.api.Assertions.fail;
 
 import java.io.IOException;
 import java.net.URI;
 
+import org.eclipse.microprofile.config.spi.ConfigSource;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
@@ -31,6 +31,7 @@ public class CodeFlowDevModeTestCase {
             UnprotectedResource.class,
             CustomTenantConfigResolver.class,
             CustomTokenStateManager.class,
+            OidcConfigSource.class,
             SecretProvider.class
     };
 
@@ -38,7 +39,8 @@ public class CodeFlowDevModeTestCase {
     static final QuarkusDevModeTest test = new QuarkusDevModeTest()
             .withApplicationRoot((jar) -> jar
                     .addClasses(testClasses)
-                    .addAsResource("application-dev-mode.properties", "application.properties"));
+                    .addAsResource("application-dev-mode.properties", "application.properties")
+                    .addAsServiceProvider(ConfigSource.class, OidcConfigSource.class));
 
     @Test
     public void testAccessAndRefreshTokenInjectionDevMode() throws IOException, InterruptedException {
@@ -47,13 +49,13 @@ public class CodeFlowDevModeTestCase {
 
         try (final WebClient webClient = createWebClient()) {
 
-            // Default tenant is disabled and client-id is wrong
+            // Default tenant is disabled and client secret is wrong
             HtmlPage page = webClient.getPage("http://localhost:8080/unprotected");
             assertEquals("unprotected", page.getBody().asText());
 
             try {
                 webClient.getPage("http://localhost:8080/protected");
-                fail("Exception is expected because the tenant is disabled and invalid client_id is used");
+                fail("Exception is expected because the tenant is disabled and invalid client secret is used");
             } catch (FailingHttpStatusCodeException ex) {
                 // Reported by Quarkus
                 assertEquals(401, ex.getStatusCode());
@@ -61,18 +63,26 @@ public class CodeFlowDevModeTestCase {
 
             // Enable the default tenant
             test.modifyResourceFile("application.properties", s -> s.replace("tenant-enabled=false", "tenant-enabled=true"));
-            // Default tenant is enabled, client-id is wrong
+            // Default tenant is enabled, client secret is wrong
             try {
-                webClient.getPage("http://localhost:8080/protected");
-                fail("Exception is expected because the tenant is disabled and invalid client_id is used");
+                page = webClient.getPage("http://localhost:8080/protected");
+
+                assertEquals("Sign in to quarkus", page.getTitleText());
+
+                HtmlForm loginForm = page.getForms().get(0);
+
+                loginForm.getInputByName("username").setValueAttribute("alice");
+                loginForm.getInputByName("password").setValueAttribute("alice");
+
+                page = loginForm.getInputByName("login").click();
+                fail("Exception is expected because an invalid client secret is used");
             } catch (FailingHttpStatusCodeException ex) {
-                // Reported by Keycloak
-                assertEquals(400, ex.getStatusCode());
-                assertTrue(ex.getResponse().getContentAsString().contains("Client not found"));
+                assertEquals(401, ex.getStatusCode());
             }
+            webClient.getCookieManager().clearCookies();
 
             // Now set the correct client-id
-            test.modifyResourceFile("application.properties", s -> s.replace("client-dev", "quarkus-web-app"));
+            test.modifyResourceFile("application.properties", s -> s.replace("secret-from-vault-typo", "secret-from-vault"));
 
             page = webClient.getPage("http://localhost:8080/protected");
 

--- a/extensions/oidc/deployment/src/test/java/io/quarkus/oidc/test/OidcConfigSource.java
+++ b/extensions/oidc/deployment/src/test/java/io/quarkus/oidc/test/OidcConfigSource.java
@@ -1,0 +1,24 @@
+package io.quarkus.oidc.test;
+
+import java.util.Set;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+public class OidcConfigSource implements ConfigSource {
+
+    @Override
+    public Set<String> getPropertyNames() {
+        return Set.of("oidc.client-id");
+    }
+
+    @Override
+    public String getValue(String propertyName) {
+        return "oidc.client-id".equals(propertyName) ? "quarkus-web-app" : null;
+    }
+
+    @Override
+    public String getName() {
+        return "oidc";
+    }
+
+}

--- a/extensions/oidc/deployment/src/test/resources/application-dev-mode.properties
+++ b/extensions/oidc/deployment/src/test/resources/application-dev-mode.properties
@@ -1,9 +1,9 @@
 quarkus.oidc.auth-server-url=${keycloak.url}/realms/quarkus
 quarkus.oidc.tenant-enabled=false
-# This is a wrong client-id, will be updated to 'quarkus-web-app' in the dev mode test
-quarkus.oidc.client-id=client-dev
+quarkus.oidc.client-id=${oidc.client-id}
 quarkus.oidc.credentials.client-secret.provider.name=vault-secret-provider
-quarkus.oidc.credentials.client-secret.provider.key=secret-from-vault
+# This is a wrong client secret key, will be updated to 'secret-from-vault' in the dev mode test
+quarkus.oidc.credentials.client-secret.provider.key=secret-from-vault-typo
 quarkus.oidc.application-type=web-app
 quarkus.oidc.logout.path=/protected/logout
 

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfigPropertySupplier.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfigPropertySupplier.java
@@ -1,0 +1,50 @@
+package io.quarkus.oidc.runtime;
+
+import java.util.function.Supplier;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+
+public class OidcConfigPropertySupplier implements Supplier<String> {
+
+    private String oidcConfigProperty;
+    private String defaultValue;
+
+    public OidcConfigPropertySupplier() {
+
+    }
+
+    public OidcConfigPropertySupplier(String oidcConfigProperty) {
+        this(oidcConfigProperty, null);
+    }
+
+    public OidcConfigPropertySupplier(String oidcConfigProperty, String defaultValue) {
+        this.oidcConfigProperty = oidcConfigProperty;
+        this.defaultValue = defaultValue;
+    }
+
+    @Override
+    public String get() {
+        if (defaultValue != null) {
+            return ConfigProvider.getConfig().getOptionalValue(oidcConfigProperty, String.class).orElse(defaultValue);
+        } else {
+            return ConfigProvider.getConfig().getValue(oidcConfigProperty, String.class);
+        }
+    }
+
+    public String getOidcConfigProperty() {
+        return oidcConfigProperty;
+    }
+
+    public void setOidcConfigProperty(String oidcConfigProperty) {
+        this.oidcConfigProperty = oidcConfigProperty;
+    }
+
+    public String getDefaultValue() {
+        return defaultValue;
+    }
+
+    public void setDefaultValue(String defaultValue) {
+        this.defaultValue = defaultValue;
+    }
+
+}

--- a/integration-tests/oidc-token-propagation/src/main/java/io/quarkus/it/keycloak/OidcConfigSource.java
+++ b/integration-tests/oidc-token-propagation/src/main/java/io/quarkus/it/keycloak/OidcConfigSource.java
@@ -1,0 +1,24 @@
+package io.quarkus.it.keycloak;
+
+import java.util.Set;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+public class OidcConfigSource implements ConfigSource {
+
+    @Override
+    public Set<String> getPropertyNames() {
+        return Set.of("oidc.client-id");
+    }
+
+    @Override
+    public String getValue(String propertyName) {
+        return "oidc.client-id".equals(propertyName) ? "quarkus-app" : null;
+    }
+
+    @Override
+    public String getName() {
+        return "oidc";
+    }
+
+}

--- a/integration-tests/oidc-token-propagation/src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.ConfigSource
+++ b/integration-tests/oidc-token-propagation/src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.ConfigSource
@@ -1,0 +1,1 @@
+io.quarkus.it.keycloak.OidcConfigSource

--- a/integration-tests/oidc-token-propagation/src/main/resources/application.properties
+++ b/integration-tests/oidc-token-propagation/src/main/resources/application.properties
@@ -1,5 +1,5 @@
 %prod.quarkus.oidc.auth-server-url=${keycloak.url}/realms/quarkus/
-quarkus.oidc.client-id=quarkus-app
+quarkus.oidc.client-id=${oidc.client-id}
 quarkus.oidc.credentials.secret=secret
 
 quarkus.keycloak.devservices.java-opts=-Dkeycloak.profile.feature.token_exchange=enabled -Dkeycloak.profile=preview

